CNDB-15527: Upgrade jvector to 4.0.0-rc.8; make FusedPQ configurable

michaeljmarshall opened the pull request:
### What is the issue

Fixes: https://github.com/riptano/cndb/issues/15527
CNDB test PR: https://github.com/riptano/cndb/pull/16797

### What does this PR fix and why was it fixed

This PR upgrades jvector, which brings several improvements. Here are the git commits brought in:

```
8b3e93cf (tag: 4.0.0-rc.8) chore: update changelog for 4.0.0-rc.8 (#627)
9d0488e5 release 4.0.0-rc.8 (#626)
570bd118 Refactor parallel writer (#608)
20c348ec Move buffer position in ByteBufferIndexWriter#writeFloats (#607)
d9ddce51 Ensure extractTrainingVectors return a list of at most MAX_PQ_TRAINING_SET_SIZE (#610)
d663b4f7 add config options for regression testing (#609)
7e493eee On-disk index cache for the Grid benchmark harness (#612)
e263cc80 Improved dataset loading; fixes, safeties, diagnostics, and better feedback (#613)
6b235ce7 bump to next SNAPSHOT (#605)
84bf5708 (tag: 4.0.0-rc.7) chore: update changelog for 4.0.0-rc.7 (#604)
fceeb885 release 4.0.0-rc.7 (#603)
51807cba add protection against bad ordinal mappings (#602)
6ca3b5e2 adding memory and disk usage stats to bench tests (#591)
a66fd914 Fix OnDiskGraphIndex#ramBytesUsed NPE (#588)
0ca5a392 Move float bulk-write into IndexWriter to enforce endianness (#577)
a6c6c09b Add diversityScoreFunctionFor to avoid creation of wrapper object (#592)
977c21d4 Relax the threshold of a flaky test related to an experimental feature (#598)
fa808d69 adding average nodes visited to benchmark tests (#552)
3bd15e70 Virtualize and Modularize DataSetLoader logic (#593)
42259e9f Speed up ivec reads by buffering (#584)
f967f1c9 virtualize DataSet (#589)
55f902f4 turn off parallel writes in grid (#582)
019a241d Parallelize graph writes (#542)
02fea879 Save allocation of a large array in PQVectors.encodeAndBuild (#574)
32a51821 javadoc for base [graph] (#548)
4eb607f8 javadoc for base [disk,exceptions] (#547)
30e8932c Enable the fused graph index  (#561)
d8848fc6 Start development on 4.0.0-rc.7-SNAPSHOT (#573)
c57f3a62 (tag: 4.0.0-rc.6) chore: update changelog for 4.0.0-rc.6 (#572)
214b7c20 release 4.0.0-rc.6 (#571)
e3686999 fix javadoc error (#570)
88669887 Ignoring testIncrementalInsertionFromOnDiskIndex_withNonIdentityOrdinalMapping and adding a TODO in buildAndMergeNewNodes (#569)
29a943e1 Computation of reconstruction errors for vector compressors (#567)
d8e9cb16 Add NVQ paper in README (#560)
d5cbe658 Add ImmutableGraphIndex.isHierarchical (#563)
b484dae2 Harden tests for heap graph reconstruction (#543)
9471c57d Make the thresholds in TestLowCardinalityFiltering tighter (#559)
21e4a226 Begin development on 4.0.0-rc.6 (#558)
4f661d99 Revert "Start development on 4.0.0-rc.6-SNAPSHOT"
fdee5779 Start development on 4.0.0-rc.6-SNAPSHOT
```

### SAI Version Bump

Adds a new sai on disk version: `fa`

### Fused PQ

With this version, we are adding a new, experimental feature to write PQ vectors fused into the graph. In doing so, we are able to skip writing the PQ vectors to the PQ file, which results in significant memory savings since the PQ vectors in the `CassandraDiskAnn` graph searcher consumers `O(n)` memory based on the number of vectors and their quantized size. The fused pq vectors mostly fit within the page cache as we read the node and its neighbors from disk, so we see minimal latency reduction due to this change, though further testing is required to see the real impact.

In order to enable fused pq, the runtime needs `cassandra.sai.latest.version=fa` or greater and `cassandra.sai.vector.enable_fused=true`. Note that because this feature is still experimental, `cassandra.sai.vector.enable_fused` defaults to `false`.

Another experimental feature introduced in this commit via the jvector upgrade is parallel graph encoding and writing to disk. Writing the fused graph requires increased CPU time to encode the graph node and we write more bytes to disk, so this parallelism is likely necessary to keep vector index creation/compaction times down. The key configurations available with their associated defaults:

```java
    // When building a compaction graph, encode layer 0 nodes in parallel and subsequently use async io for writes.
    // This feature is experimental, so defaults to false.
    SAI_ENCODE_AND_WRITE_VECTOR_GRAPH_IN_PARALLEL_ENABLED("cassandra.sai.vector.encode_and_write_graph_in_parallel.enabled", "false"),
    // When parallel graph encoding is enabled, the number of threads to use for encoding. Defaults to 0, meaning
    // use all available processors as reported by the JVM.
    SAI_ENCODE_AND_WRITE_VECTOR_GRAPH_IN_PARALLEL_NUM_THREADS("cassandra.sai.vector.encode_and_write_graph_in_parallel.num_threads", "0"),
    // When parallel graph encoding is enabled, whether to use director buffers. Defaults to false, meaning heap
    // buffers are used. A buffer will be allocated per encoding thread. The size of each buffer is the size
    // of the encoded graph node at layer 0, which varies based on graph feature settings.
    SAI_ENCODE_AND_WRITE_VECTOR_GRAPH_IN_PARALLEL_USE_DIRECT_BUFFERS("cassandra.sai.vector.encode_and_write_graph_in_parallel.use_direct_buffers", "false"),
```

### OnDiskVectorValues and OnDiskVectorValuesWriter

`OnDiskVectorValues` is now in its own file and is now thread safe in order to account for some necessary implementation details within jvector. Added `OnDiskVectorValuesWriter` to improve test coverage and to abstract away the flush issues associated with `BufferedRandomAccessWriter` as described in https://github.com/datastax/jvector/issues/562.

### Verification

This PR also introduces new benchmarks as well as improved unit testing. The new benchmarks verify the performance of the `OnDiskVectorValues` and `OnDiskVectorValuesWriter` to confirm (at least directionally) the time associated with read and write operations.

New tests have been added to verify that when we iterate over an sstable's rows, we are able to assert that the sstable's vector value's similarity to the one stored in the vector graph is ~1. This testing is valuable in that it confirms the row id to ordinal mapping is correct at every node. Previously, we relied on recall results to verify this for us. This new pattern allows us to confirm _every_ node, which is more thorough and removes most edge cases that might have led to partially correct graphs that may have achieved acceptable recall.

michaeljmarshall committed:
CNDB-15554: Bump jvector version

michaeljmarshall committed:
Work in progress for fused adc, fails when querying compacted graph

michaeljmarshall committed:
Skip writing fused graph if version to early

michaeljmarshall committed:
Comment out writeHeader to fix build

michaeljmarshall committed:
Merge branch 'cndb-15554' into revive-fused-adc

michaeljmarshall committed:
Add writeHeader back in to fix failing compaction test

michaeljmarshall committed:
CNDB-15640: Determine if vectors are unit length at insert

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Remove unused variable

michaeljmarshall committed:
Merge branch 'cndb-15640' into revive-fused-adc

michaeljmarshall committed:
Save progress on COHG (doesn't work yet)

michaeljmarshall committed:
Save progress (works now, but need to consider refactoring)

michaeljmarshall committed:
Fix writing header before closing CompactionGraph's first ODGIW

michaeljmarshall commented on src/java/org/apache/cassandra/index/sai/disk/vector/CassandraDiskAnn.java:
```diff
@@ -117,7 +117,7 @@ public CassandraDiskAnn(SSTableContext sstableContext, SegmentMetadata.Component
 
             VectorCompression.CompressionType compressionType = VectorCompression.CompressionType.values()[reader.readByte()];
-            if (features.contains(FeatureId.FUSED_ADC))
+            if (features.contains(FeatureId.FUSED_PQ))
```
@marianotepper - I just noticed that the `features` map already has logic that loads the `ProductQuantization`, meaning this branch currently keeps two identical maps in memory. I think it'd make sense to possibly expose the `features` map in the `OnDiskGraphIndex` so we can remove the duplicate cost. Any reason we can't do that?

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Update for latest code changes

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Use internal build of jvector

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Bring in jvector bug fix

built from the branch in this
PR: https://github.com/datastax/jvector/pull/588

michaeljmarshall committed:
Add temp hack to parallelize pq computation

This is a hack because it increases memory
utilization and may impact other threads
in the CC host. However, this will help
unblock certain test scenarios.

eolivelli committed:
CNDB-16243: Improve ChunkCache Key in order to keep O(1) access time

Improve hashCode of the Key to prevent collisions
Implement Comparable<Key> in case the ConcurrentHashMap falls back in to Tree mode

(cherry picked from commit 091b6363959dbc9f879e31859c77c33d254856d7)

eolivelli committed:
CNDB-16252: temporary hack to allow building multiple vector indexes per table

michaeljmarshall committed:
CNDB-16308: Add GLOBAL_HOLES_ALLOWED logic to CompactionGraph

michaeljmarshall committed:
Fix indentation

michaeljmarshall committed:
Add cassandra.sai.vector.encode_write_graph_parallel, default to true

michaeljmarshall committed:
Pass local path to OnDiskGraphIndexWriter for parallel writes

michaeljmarshall committed:
Revert "CNDB-16243: Improve ChunkCache Key in order to keep O(1) access time"

We'll pull this in when we merge from main.

This reverts commit 1d9129c249ed3077fd2cdbd11af7286a22dd84d1.

michaeljmarshall committed:
Revert "CNDB-16252: temporary hack to allow building multiple vector indexes per table"

This reverts commit 29b6121e3bcc1612260ea2cdd31980900e77ee19.

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Bump to jvector 4.0.0-rc.7

michaeljmarshall committed:
Make SequentialWriter implement jvector IndexWriter

michaeljmarshall committed:
Bump to feature branch build of jvector to run all SAI tests

michaeljmarshall committed:
Revert "Bump to feature branch build of jvector to run all SAI tests"

The feature branch build worked (passed tests) but
it was only needed for parallel index writes, which
are apparently not prod ready, so going to
remove that config for now and go back to the
latest official jvector release

This reverts commit e46d642f0c406cf0599beb884a59907f84d3fac9.

michaeljmarshall committed:
Disable encode_write_graph_parallel by default for now

michaeljmarshall committed:
Revert "CNDB-16308: Add GLOBAL_HOLES_ALLOWED logic to CompactionGraph"

Comes in via: https://github.com/datastax/cassandra/pull/2203
(this commit alone was actually wrong...)

This reverts commit 213a2b9ba5b09bf657c1e367f0396d92f01ac5b6.

michaeljmarshall committed:
Fix license header; linearize parallel PQ encoding; modify comment

michaeljmarshall committed:
Use new writeFloats method for optimized vector writes

michaeljmarshall committed:
Work around jvector issue 562 for better buffering

michaeljmarshall committed:
Refactor flat vector writer/reader into classes

This commit primarily focuses on increasing
test coverage for this low level
component where we write vectors by their
ordinal position. The added test coverage
was generated by Bob.

michaeljmarshall committed:
Empty commit to trigger CI

michaeljmarshall committed:
Empty commit to trigger CI

pkolaczk commented on src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java:
```diff
@@ -449,36 +460,24 @@ public SegmentMetadata.ComponentMetadataMap flush(IndexComponents.ForWrite perIn
-            long pqPosition = writePQ(pqOutput.asSequentialWriter(), remappedPostings, perIndexComponents.context());
-            long pqLength = pqPosition - pqOffset;
+            var compressor = writePQ(pqOutput.asSequentialWriter(), remappedPostings, perIndexComponents.context(), writeFusedPQ);
+            long pqLength = pqOutput.asSequentialWriter().position() - pqOffset;
```
nit: can use `pqOutput.getFilePointer()`

pkolaczk commented on src/java/org/apache/cassandra/index/sai/disk/vector/VectorByOrdinalWriter.java:
```diff
@@ -0,0 +1,123 @@
+ * - Efficient buffering via BufferedRandomAccessWriter
+ * This class is not thread-safe and should only be used within the vector index package.
+ */
+class VectorByOrdinalWriter implements Closeable
```
Isn't it the writer for the file read by `OnDiskVectorValues` ?
If so, shouldn't it be named `OnDiskVectorValuesWriter` instead? 
A different name suggests it is something else.

pkolaczk reviewed (commented):
I'm not knowledgeable enough to give a binding +1 for this, but overall looks good and I can see the code is cleaner than it was before. 

One important question I have is - how do we actually test it? I can't find any changes to tests related to FUSED ADC actually doing anything. Like, I can see it's enabled by default now, so we're probably testing it already with existing tests; right? But those tests were passing with fused adc disabled before. Maybe we should add at least a smoke test that would verify that when. fused adc is enabled, it really does something (I guess - saves memory, right?).

michaeljmarshall committed:
Address initial review comments

michaeljmarshall committed:
Rename VectorByOrdinalWriter => OnDiskVectorValuesWriter

michaeljmarshall committed:
Fix CNDB-15926

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Disable fused by default; refactor to simplify test parameterization

michaeljmarshall commented:
@pkolaczk - the feature is now disabled by default. It is still covered by tests, but we will merge this for now with the caveat that we won't enable fused until we have increased test coverage.

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Remove unnecessary reference to binary quantization

With this work https://github.com/riptano/cndb/issues/16693
in flight in parallel, it only makes sense
to remove this here now. This code
is unreachable at the moment anyway.

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

michaeljmarshall committed:
Remove the double compact in testCompaction()

One of the commits from main covered
this test case better than calling
compact twice in a row, so this
commit backs out the original
change to double compact.

michaeljmarshall committed:
Bump to jvector 4.0.0-rc.8

michaeljmarshall committed:
Remove blank comment line

michaeljmarshall committed:
Add OnDiskVectorValues and OnDiskVectorValuesWriter benchmark

     [java] Benchmark                                         (dimension)  Mode  Cnt    Score    Error  Units
     [java] OnDiskVectorValuesBench.readAllVectorsSequential          384  avgt    5   55.764 ±  1.522  ms/op
     [java] OnDiskVectorValuesBench.readAllVectorsSequential         1536  avgt    5  198.640 ±  3.520  ms/op
     [java] OnDiskVectorValuesBench.writeVectorsToDisk                384  avgt    5  113.735 ±  2.768  ms/op
     [java] OnDiskVectorValuesBench.writeVectorsToDisk               1536  avgt    5  500.743 ± 18.353  ms/op

michaeljmarshall committed:
Refactor OnDiskVectorValues to use shared array given usage patterns allow it

Note that the main motivation for updating this
logic comes from the implementation for the
RandomAccessVectorValues method named
threadLocalSupplier, which makes an inference
based on the result of isValueShared about
whether the class instance can be shared across
threads. Since this one cannot, it is safest
to just make the value shared. Given
how we use the resulting vector at the moment,
it is not a problem to make it shared.

     [java] Benchmark                                         (dimension)  Mode  Cnt    Score    Error  Units
     [java] OnDiskVectorValuesBench.readAllVectorsSequential          384  avgt    5   52.610 ±  2.211  ms/op
     [java] OnDiskVectorValuesBench.readAllVectorsSequential         1536  avgt    5  184.552 ±  4.006  ms/op
     [java] OnDiskVectorValuesBench.writeVectorsToDisk                384  avgt    5  113.262 ±  3.906  ms/op
     [java] OnDiskVectorValuesBench.writeVectorsToDisk               1536  avgt    5  498.166 ± 14.529  ms/op

michaeljmarshall committed:
Clean up some boolean logic

michaeljmarshall committed:
Refactor implementation again to be thread safe

Note that the score is slightly worse, but the
error ratio is high. There is certainly cost
to fetching the thread from the hash map.

The main motivation here is to prevent leaked
resources. See https://github.com/datastax/jvector/issues/635

     [java] Benchmark                                         (dimension)  Mode  Cnt    Score    Error  Units
     [java] OnDiskVectorValuesBench.readAllVectorsSequential          384  avgt    5   59.925 ±  4.414  ms/op
     [java] OnDiskVectorValuesBench.readAllVectorsSequential         1536  avgt    5  212.512 ± 28.455  ms/op
     [java] OnDiskVectorValuesBench.writeVectorsToDisk                384  avgt    5  117.685 ±  4.208  ms/op
     [java] OnDiskVectorValuesBench.writeVectorsToDisk               1536  avgt    5  526.014 ± 14.489  ms/op

michaeljmarshall committed:
Verify rowid to ordinal mapping correctness within graph

michaeljmarshall committed:
Cover the missing vector case

michaeljmarshall committed:
Fix assertions to account for rows with missing vectors outside segment bounds

michaeljmarshall committed:
Commit the benchmark referenced in earlier commits

I ran it again, and the results seem
in the same ballpark, but with higher
error bounds.

     [java] Benchmark                                         (dimension)  Mode  Cnt    Score     Error  Units
     [java] OnDiskVectorValuesBench.readAllVectorsSequential          384  avgt    5   61.719 ±  17.843  ms/op
     [java] OnDiskVectorValuesBench.readAllVectorsSequential         1536  avgt    5  217.288 ±  37.908  ms/op
     [java] OnDiskVectorValuesBench.writeVectorsToDisk                384  avgt    5  117.555 ±   5.090  ms/op
     [java] OnDiskVectorValuesBench.writeVectorsToDisk               1536  avgt    5  560.162 ± 251.997  ms/op

michaeljmarshall committed:
Adjust OnDiskVectorValuesTest assertions to match updated thread safe design

michaeljmarshall committed:
Fix failing test related to posting structure

michaeljmarshall committed:
Tweak compaction test posting structure again

michaeljmarshall committed:
Merge remote-tracking branch 'datastax/main' into revive-fused-adc

MarkWolters commented on src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java:
```diff
@@ -496,25 +490,50 @@ public SegmentMetadata.ComponentMetadataMap flush(IndexComponents.ForWrite perIn
+            .with(nvq != null ? new NVQ(nvq) : new InlineVectors(vectorValues.dimension()));
+
+        if (compressor instanceof ProductQuantization && JVectorVersionUtil.shouldWriteFused(context.version()))
+            indexWriterBuilder.with(new FusedPQ(context.getIndexWriterConfig().getAnnMaxDegree(), (ProductQuantization) compressor));
```
As far as I can tell compressor is always going to be either null or an instance of ProductQuantization here. It is instantiated on line 473 by calling writePQ, which either returns a ProductQuantization or null. Here there is no null check and it's also being checked and cast to ProductQuantization which seems dangerous.

ekaterinadimitrova2 replied:
Will take a look, thanks!

ekaterinadimitrova2 replied:
Line 531:
`null instanceof ProductQuantization -> returns false`, so I think we are good. To make it more clear, we can add some assertions maybe?

MarkWolters commented on src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java:
```diff
@@ -496,25 +490,50 @@ public SegmentMetadata.ComponentMetadataMap flush(IndexComponents.ForWrite perIn
         }
     }
 
+    private OnDiskGraphIndexWriter createIndexWriter(File indexFile, long termsOffset, IndexContext context, OrdinalMapper ordinalMapper, VectorCompressor<?> compressor, NVQuantization nvq) throws IOException
```
CassandraRelevantProperties implements configuration to allow for using parallel graph writes but here they are not referenced and the serial writer is instantiated.

ekaterinadimitrova2 replied:
Good point. I was reading in the PR:

> Another experimental feature introduced in this commit via the jvector upgrade is parallel graph encoding and writing to disk. Writing the fused graph requires increased CPU time to encode the graph node and we write more bytes to disk, so this parallelism is likely necessary to keep vector index creation/compaction times down. The key configurations available with their associated defaults:

From `Writing the fused graph requires increased CPU time` I understand we would want to enable this config. I have a note while looking into the testing also to check where it is tested. Thanks for flagging!

ekaterinadimitrova2 replied:
@MarkWolters , are the parallel graph writes still experimental as the comments say, or they are in the same boat as FusedPQ - your test framework was fixed and you got good results for both and it is ok to potentially enable them by default after we finalize testing on your end too?

MarkWolters replied:
@ekaterinadimitrova2 I would categorize them as still experimental. The feature has been tested internally but not to the extent of fused PQ. I believe it is ready to use, I'd just feel more comfortable once it goes through some testing once integrated to validate the correctness of the written graphs.

ekaterinadimitrova2 replied:
I see, 

>  I'd just feel more comfortable once it goes through some testing once integrated

You mean in Cassandra/Astra I guess and not JVector additional testing you plan? Did I get it right?

ekaterinadimitrova2 replied:
We use the parallelism in the `CompactionGraph` and it is everywhere mentioned compaction.
https://github.com/datastax/cassandra/blob/522e8b62875e6ba049647c87efa5c9915e54b098/src/java/org/apache/cassandra/index/sai/disk/vector/CompactionGraph.java#L238-L249

I think it is actually intentionally not considered here in flushing to use parallelism which was meant to address disk I/I bottleneck. @pkolaczk , do you remember this from your review?

pkolaczk replied:
Seriously, I don't recall having any conversation about that. Is the parallel writing using more memory than sequential? I understand that FusedPQ requires a bit more work when encoding / building the graph. 
But also, when flushing, maybe we already have the PQ encoded and there is no need to go parallel, because there is less work to do? (just guessing, will check soon)

ekaterinadimitrova2 replied:
I think you are right about the encoding. 
`CassandraOnHeapGraph` (Memtable Flush):

- Often reuses existing PQ from previous segments (line 642)
- When writing FusedPQ, skips separate PQ encoding (line 647: !writeFusedPQ)

The graph write itself becomes the main work, which may not parallelize well with in-memory data?

ekaterinadimitrova2 replied:
I also found this comment - https://github.com/datastax/cassandra/blob/522e8b62875e6ba049647c87efa5c9915e54b098/src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java#L635-L636

pkolaczk replied:
There is this comment here that might shed some light on parallel writing:
https://github.com/datastax/cassandra/blob/522e8b62875e6ba049647c87efa5c9915e54b098/src/java/org/apache/cassandra/index/sai/disk/vector/CassandraOnHeapGraph.java#L635-L636

Looks like it was meant to be parallel, but one index at a time.

ekaterinadimitrova2 replied:
Yup, same one I linked but still not convinced what is right

ekaterinadimitrova2 replied:
Talked to Michael, this is intentional. We do not have FusedPQ wired on flush, only on compaction and that is OK as we quickly compact. Flush is already a bit of a bottleneck, so he did not want to add more work there. Any memory benefits are realized after compaction.

MarkWolters commented on test/microbench/org/apache/cassandra/test/microbench/index/sai/disk/vector/OnDiskVectorValuesBench.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    @TearDown(Level.Trial)
+    public void tearDown()
```
Why 2 teardowns with different capitalization?

ekaterinadimitrova2 replied:
It seems we are allowed but there may be a chance for them to run not in the sequence we want them. Looking into that. Maybe we can just merge them

MarkWolters commented on test/unit/org/apache/cassandra/index/sai/cql/VectorSiftSmallTest.java:
```diff
@@ -387,7 +387,7 @@ private void createTable()
     {
         // we need a long timeout because we are adding many vectors
-        String index = createIndexAsync("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex' WITH OPTIONS = {'similarity_function' : 'euclidean'}");
+        String index = createIndexAsync("CREATE CUSTOM INDEX ON %s(val) USING 'StorageAttachedIndex' WITH OPTIONS = {'similarity_function' : 'euclidean', 'enable_hierarchy': 'true'}");
```
For the future I think it would be more thorough coverage to test both with hierarchy and without

MarkWolters commented on test/unit/org/apache/cassandra/index/sai/disk/vector/BruteForceRowIdIteratorTest.java:
```diff
@@ -95,6 +96,12 @@ public NodesIterator getNeighborsIterator(int i, int i1)
 
+        @Override
+        public void processNeighbors(int i, int i1, ScoreFunction scoreFunction, ImmutableGraphIndex.IntMarker intMarker, ImmutableGraphIndex.NeighborProcessor neighborProcessor)
+        {
```
Is this intentionally empty? If so should there be a comment to that effect?

ekaterinadimitrova2 replied:
We can throw again unsupported, there was a comment for all these methods that they are not used in this test override, but better to be more clear and consistent. I will add it.

MarkWolters reviewed (commented):
Overall looks good, excited to see these features being put into use.
